libs/ml/include/ml/layers/fully_connected.hpp

```diff
@@ -46,16 +46,101 @@ class FullyConnected : public SubGraph<T>
+
+    // share weight with target_node
+    ShareWeights(target_node_ptr);
+  }
 
   FullyConnected(SizeType in, SizeType out,
                  details::ActivationType activation_type = details::ActivationType::NOTHING,
                  fetch::ml::details::RegularisationType regulariser =
                      fetch::ml::details::RegularisationType::NONE,
-                 DataType           regularisation_rate = static_cast<DataType>(0),
-                 std::string const &name = "FC", WeightsInit init_mode = WeightsInit::XAVIER_GLOROT)
+                 DataType    regularisation_rate = static_cast<DataType>(0),
+                 WeightsInit init_mode           = WeightsInit::XAVIER_GLOROT)
     : in_size_(in)
     , out_size_(out)
   {
+    // setup overall architecture of the model
+    SetupArchitecture(activation_type, regulariser, regularisation_rate);
+
+    // initialize the weights and bias with specified initialization method
+    InitializeWeights(init_mode);
+  }
+
+  void InitializeWeights(WeightsInit init_mode)
+  {
+    // initialize weight with specified method
+    std::string name = DESCRIPTOR;
+    ArrayType   weights_data(std::vector<SizeType>({out_size_, in_size_}));
+    this->Initialise(weights_data, init_mode);
+    this->SetInput(name + "_Weights", weights_data);
+    ArrayType bias_data(std::vector<SizeType>({out_size_, 1}));
+    this->SetInput(name + "_Bias", bias_data);
+  }
+
+  void ShareWeights(NodePtrType target_node_ptr)
+  {
+    // Get ptr to each weights layer
+    std::string    name                    = DESCRIPTOR;
+    GraphPtrType   target_graph_ptr        = std::dynamic_pointer_cast<GraphType>(target_node_ptr);
+    NodePtrType    target_weights_node_ptr = target_graph_ptr->GetNode(name + "_Weights");
+    WeightsPtrType target_weights_ptr =
+        std::dynamic_pointer_cast<WeightsType>(target_weights_node_ptr);
+    NodePtrType    target_bias_node_ptr = target_graph_ptr->GetNode(name + "_Bias");
+    WeightsPtrType target_bias_ptr  = std::dynamic_pointer_cast<WeightsType>(target_bias_node_ptr);
+    NodePtrType    weights_node_ptr = this->GetNode(name + "_Weights");
+    WeightsPtrType weights_ptr      = std::dynamic_pointer_cast<WeightsType>(weights_node_ptr);
+    NodePtrType    bias_node_ptr    = this->GetNode(name + "_Bias");
+    WeightsPtrType bias_ptr         = std::dynamic_pointer_cast<WeightsType>(bias_node_ptr);
+
+    // check if the shared weight is compatible with input shape
+    auto w_s = target_weights_ptr->get_weights().shape();
+    auto b_s = target_bias_ptr->get_weights().shape();
+    if ((w_s[0] != out_size_) || (w_s[1] != in_size_) || (b_s[0] != out_size_))
+    {
+      throw std::runtime_error("the shared weight from node: " + target_node_ptr->GetNodeName() +
+                               " is incompatible with the specified shape of this layer");
+    }
+
+    // Set sudo weights to handle initilization issues
```

Comment:
- Isn't there a way around this? This is wasteful to create irrelevant tensors
- Also you mean "pseudo" not sudo (which mean superuser do)

Reply:
haha u caught me. tried to be lazy lol as this wont really affect efficiency given it happens only at initialization.